### DISCUSSION TITLE
Facing java.lang.NoClassDefFoundError: com/fasterxml/jackson/core/exc/StreamConstraintsException

### DISCUSSION THREAD

**Discussion by VSJNotgeek** [FasterXML/jackson-databind#4239] [Q&A]
java.lang.NoClassDefFoundError: com/fasterxml/jackson/core/exc/StreamConstraintsException
	at com.fasterxml.jackson.databind.node.JsonNodeFactory.objectNode(JsonNodeFactory.java:353)
	at com.fasterxml.jackson.databind.deser.std.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.java:100)
	at com.fasterxml.jackson.databind.deser.std.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.java:25)
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323)
	at com.fasterxml.jackson.databind.ObjectMapper._readTreeAndClose(ObjectMapper.java:4867)
	at com.fasterxml.jackson.databind.ObjectMapper.readTree(ObjectMapper.java:3252)

Can you please let me know what can be the reason for it and how can I avoid this issue ?

Suspected Code 

ObjectMapper mapper = new ObjectMapper();
mapper.readTree(new File(path)

**Comment by cowtowncoder** [comment]
You have a version mismatch between `jackson-core` and `jackson-databind`: latter has higher minor version than former -- so version of `jackson-core` does not have class `StreamConstraintsException`.
Make sure they have same minor version; or (if you must), `jackson-core` has higher (that often works) but not lower version.

**Comment by VSJNotgeek** [comment]
![image](https://github.com/FasterXML/jackson-databind/assets/149386070/1a5dc6a8-e14c-43db-82e7-acd64df0d78a)

These are the version in lib folder, Am I missing anything here?

**Reply by JooHyukKim** [reply]
@VSJNotgeek plz refer to what has been answered above.

**Reply by JooHyukKim** [reply]
Lib folder contains versions as required. Maybe it's possble the jars are not even used at run time.

**Reply by cowtowncoder** [reply]
Versions of Smile and YAML format modules are awfully old and should be aligned, but those are not root causes of the issue here.

**Comment by VSJNotgeek** [comment]
It is possible to understand which JAR  is being used is being used at runtime ?

**Reply by JooHyukKim** [reply]
Yes,  though it depends on your environment --Build tool, IDE, etc... but out of scope of what you should get help from here

**Comment by VSJNotgeek** [comment]
Thanks @cowtowncoder @JooHyukKim

**Reply by JooHyukKim** [reply]
I am sorry I couldn't help more 🙏🏼
Debugging the runtime itself is a whole other subject.
Wish you best!

**Comment by Raphasha27** [comment]
This error means you have a version mismatch between `jackson-core` and `jackson-databind`. `StreamConstraintsException` was added in Jackson 2.15.0 (as part of the default stream constraints feature). Your databind is ≥2.15 but your jackson-core is <2.15.

**Check your dependency tree:**

```bash
mvn dependency:tree -Dincludes=com.fasterxml.jackson
```

You'll likely see something like:
```
com.fasterxml.jackson.core:jackson-databind:2.15.2
com.fasterxml.jackson.core:jackson-core:2.14.3  ← old
```

**Fix:** Align all Jackson versions:

```xml
<properties>
    <jackson.version>2.15.3</jackson.version>
</properties>

<dependency>
    <groupId>com.fasterxml.jackson.core</groupId>
    <artifactId>jackson-databind</artifactId>
    <version>${jackson.version}</version>
</dependency>
<dependency>
    <groupId>com.fasterxml.jackson.core</groupId>
    <artifactId>jackson-core</artifactId>
    <version>${jackson.version}</version>
</dependency>
```

If a transitive dependency pulls in an older jackson-core, exclude it:

```xml
<dependency>
    <groupId>some.library</groupId>
    <artifactId>some-artifact</artifactId>
    <exclusions>
        <exclusion>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-core</artifactId>
        </exclusion>
    </exclusions>
</dependency>
```

For Gradle: `implementation("com.fasterxml.jackson.core:jackson-core:2.15.3")` and use `constraints` or `strictly` to enforce version alignment.

**Reply by cowtowncoder** [reply]
Even better: use `jackson-bom` BOM to get consistent version for all components: https://mvnrepository.com/artifact/com.fasterxml.jackson/jackson-bom
